server/jetstream_api.go

```diff
@@ -1012,30 +1012,121 @@ func (s *Server) sendAPIErrResponse(ci *ClientInfo, acc *Account, subject, reply
 
 const errRespDelay = 500 * time.Millisecond
 
-func (s *Server) sendDelayedAPIErrResponse(ci *ClientInfo, acc *Account, subject, reply, request, response string, rg *raftGroup) {
-	js := s.getJetStream()
-	if js == nil {
+type delayedAPIResponse struct {
+	ci       *ClientInfo
+	acc      *Account
+	subject  string
+	reply    string
+	request  string
+	response string
+	rg       *raftGroup
+	deadline time.Time
+	next     *delayedAPIResponse
+}
+
+// Add `r` in the list that is maintained ordered by the `delayedAPIResponse.deadline` time.
+func addDelayedResponse(head, tail **delayedAPIResponse, r *delayedAPIResponse) {
+	// Check if list empty.
+	if *head == nil {
+		*head, *tail = r, r
 		return
 	}
-	var quitCh <-chan struct{}
-	js.mu.RLock()
-	if rg != nil && rg.node != nil {
-		quitCh = rg.node.QuitC()
+	// Check if it should be added at the end, which is if after or equal to the tail.
+	if r.deadline.After((*tail).deadline) || r.deadline.Equal((*tail).deadline) {
+		(*tail).next, *tail = r, r
+		return
 	}
-	js.mu.RUnlock()
+	// Find its spot in the list.
+	var prev *delayedAPIResponse
+	for c := *head; c != nil; c = c.next {
+		// We insert only if we are stricly before the current `c`.
+		if r.deadline.Before(c.deadline) {
+			r.next = c
+			if prev != nil {
+				prev.next = r
+			} else {
+				*head = r
+			}
+			return
+		}
+		prev = c
+	}
+}
 
-	s.startGoRoutine(func() {
-		defer s.grWG.Done()
+func (s *Server) delayedAPIResponder() {
+	defer s.grWG.Done()
+	var (
+		rgQuitCh   <-chan struct{}
+		head, tail *delayedAPIResponse
+		r          *delayedAPIResponse
+		tm         = time.NewTimer(time.Hour)
+	)
+	setup := func() {
```

Comment:
Minor nit but I suggest we call this function `next` instead of `setup`, and perhaps add a comment that `r` is the next response we're expecting the timer to fire for, updated only by calling `next` again.

Reply:
In fact I've just done this myself in a new commit.

Reply:
In one of my attempts, I had named it `next()`, so I would agree with that ;-).